BLSQ/iaso
# IA-2468: make pipeline code generic

Make code used for updating lqas via openhexa pipeline reusable

Related JIRA tickets : IA-2468, POLIO-1401

## Self proofreading checklist


## Doc

https://wiki.bluesquare.org/en/Dev-team/product-management/iaso/Polio/refresh_lqas_reference

## Changes

- Add `ExternalTaskModelViewSet`, a subclass of iaso's `ModelViewSet`with all the pipeline related code from `RefreshLQASData` made more generic
- Add `ExternalTaskPostSerializer` to handle external task creation
- Move `ExternalTaskSerializer` iaso.api.tasks
- Refactor `RefreshLQASData` to subclass `ExternalTaslModelViewSet`, with some overrides in validation to take into account the specificities of the LQAS use case.
- Update front-end to use the modified API
- Refactor `test_refresh_lqas`

**Note**: `test_refresh_lqas` actually tests the external task's flow. It should also be moved and renamed, but first we need to move the `Config` object out of the polio plugin, which is another ticket: [POLIO_1439](https://bluesquare.atlassian.net/browse/POLIO-1439). Moving the tests will be done as part of that ticket

## How to test

I have tested the feature on staging (see screenshots)

To test locally:
- Set up openhexa local testing:https://github.com/BLSQ/iaso-pipelines/wiki/Setup-Iaso-to-test-OpenHexa-Polio-integration
- Go to django Admin>JsonDatastores
- Chose a an lqas  datastore
- Chose a campaign. Delete some district data
- Go to polio>lqas>country
- Select a campaign and click refresh:
    - The button should enter loading state and the last run status should be "Running" 
    - Go to Tasks: you should see the task running with the name: "lqas-pipeline-config-<countryID>"
    - Go to OpenHexa: you should see a pipeline running. Open the pipeline, it should have the same taskId as the task in iaso/polio
- Once the pipeline stops running (can take several miniutes, it's a separate issue):
    - Check in Openhexa that there were no errors, in particular no errors coming from iaso when:
        - Trying to write in the datastore
        - Trying to update the task status
        - Trying to write in the campaign 
    - Check in iaso:
        - Task status should be updated
        - The map should refresh without needing to reload 

## Print screen / video

The (renamed) task ran successfully (staging)
![Screenshot 2024-01-23 at 15 00 21](https://github.com/BLSQ/iaso/assets/38907762/cd629cb4-54f6-47d8-910a-f3769547efff)

It launched the pipeline successfully (statging)
![Screenshot 2024-01-23 at 15 00 36](https://github.com/BLSQ/iaso/assets/38907762/b2cbece1-1a75-45fd-89c4-69618648ec3c)

The data was refreshed (see status upper right)
![Screenshot 2024-01-23 at 15 00 51](https://github.com/BLSQ/iaso/assets/38907762/18feabc8-4146-4fa6-b001-c69d20c68bb6)

## Note 

Merges in https://github.com/BLSQ/iaso/pull/1062

The doc is a bit vague on some implementation details because I wanted it to fit the current implementation as well

Better not to merge before [POLIO_1439](https://bluesquare.atlassian.net/browse/POLIO-1439) is done, since the feature won't work outside of polio
